src/experimental/tools/videoThumbnailLoader/init_source_buffer.ts

```diff
@@ -0,0 +1,150 @@
+} from "rxjs/operators";
+import createSegmentLoader from "../../../core/pipelines/segment/create_segment_loader";
+import { QueuedSourceBuffer } from "../../../core/source_buffers";
+import dash from "../../../transports/dash";
+import prepareSourceBuffer from "./prepare_source_buffer";
+import { IContentInfos } from "./types";
+
+const _currentVideoSourceBuffers: WeakMap<HTMLMediaElement,
+                                          QueuedSourceBuffer<Uint8Array>> = new WeakMap();
+const _currentContentInfos: WeakMap<HTMLMediaElement,
+                                    IContentInfos> = new WeakMap();
+
+const { loader, parser } = dash({ lowLatencyMode: false }).video;
+const segmentLoader = createSegmentLoader(
+  loader,
+  { maxRetry: 0,
+    maxRetryOffline: 0,
+    initialBackoffDelay: 0,
+    maximumBackoffDelay: 0, });
+
+/**
+ * Get current source buffer :
+ * - If already created for current representation, reuse
+ * - If new codecs and/or init URL, create a new one
+ * @param {Object} contentInfos
+ * @param {HTMLVideoElement} element
+ * @returns {Observable}
+ */
+export function initSourceBuffer$(contentInfos: IContentInfos,
+                                  element: HTMLVideoElement
+): Observable<QueuedSourceBuffer<Uint8Array>> {
+  let _sourceBufferObservable$: Observable<QueuedSourceBuffer<Uint8Array>>;
+  const currentVideoSourceBuffer = _currentVideoSourceBuffers.get(element);
+  const currentContentInfos = _currentContentInfos.get(element);
+  if (currentContentInfos !== undefined &&
+      currentVideoSourceBuffer !== undefined) {
+    _sourceBufferObservable$ = observableOf(currentVideoSourceBuffer);
+  } else {
+    _sourceBufferObservable$ =
+      prepareSourceBuffer(element,
+                          contentInfos.representation.getMimeTypeString())
+        .pipe(
+          map((videoSourceBuffer) => {
+            currentVideoSourceBuffer?.dispose();
+            _currentVideoSourceBuffers.set(element, videoSourceBuffer);
+            return videoSourceBuffer;
+          }),
+          catchError((err: Error) => {
+            throw new Error("VideoThumbnailLoaderError: Error when creating" +
+                            " media source or source buffer: " + err.toString());
+          }),
+          shareReplay()
+        );
+  }
+
+  const initSegment = contentInfos.representation.index.getInitSegment();
+  const currentInitSegmentId =
+    currentContentInfos?.representation.index.getInitSegment()?.id;
+
+  if (currentInitSegmentId === initSegment?.id &&
+      contentInfos.representation.id === currentContentInfos?.representation.id &&
+      contentInfos.adaptation.id === currentContentInfos?.adaptation.id &&
+      contentInfos.period.id === currentContentInfos?.period.id &&
+      contentInfos.manifest.id === currentContentInfos?.manifest.id) {
+    return _sourceBufferObservable$;
+  }
+
+  return _sourceBufferObservable$.pipe(
+    mergeMap((sourceBuffer) => {
+      if (initSegment == null) {
```

Comment:
We could just do without no?

Reply:
(Not that it is possible today, but it could be)

Reply:
still true

Reply:
Hmm what do you mean ? The init segment is still needed for buffering content, no ?

Reply:
There could be cases where it is included in all regular segments.

In that case, the init segment won't be needed.

I think (not sure) that this is the case with mpeg2ts